dali/kernels/signal/fft/fft_postprocess.cuh

```diff
@@ -0,0 +1,385 @@
+
+    if (in.is_contiguous() && out.is_contiguous()) {
+      int64_t nwindows = 0;
+      for (int i = 0; i < N; i++) {
+        nwindows += in.shape[i][0];
+      }
+
+      dim3 blocks(1, div_ceil(nwindows, kBlock));
+      dim3 threads(32, kBlock);
+
+      ConvertTimeMajorSpectrogram<<<blocks, threads, 0, ctx.gpu.stream>>>(
+          out.data[0], out.shape[0][1], in.data[0], in.shape[0][1], nfft, nwindows, convert_
+        );
+      CUDA_CALL(cudaGetLastError());
+    } else {
+      for (int i = 0; i < N; i++) {
+        int nwindows = in.shape[i][0];
+
+        dim3 blocks(1, div_ceil(nwindows, kBlock));
+        dim3 threads(32, kBlock);
+
+        ConvertTimeMajorSpectrogram<<<blocks, threads, 0, ctx.gpu.stream>>>(
+            out.data[i], out.shape[i][1], in.data[i], in.shape[i][1], nfft, nwindows, convert_
+          );
+        CUDA_CALL(cudaGetLastError());
+      }
+    }
+  }
+
+ private:
+  Convert convert_;
+};
+
+/**
+ * A specialized kernel that tranposes a frame-major spectrogrum to frequency-major one,
+ * with an option to apply some pointwise transform on the data (e.g. complex magnitude).
+ *
+ * Constraints: all input samples must have the same
+ */
+template <typename Out, typename In = Out, typename Convert = identity>
+class ToFreqMajorSpectrum : public FFTPostprocess<Out, In> {
+ public:
+  using SampleDesc = fft_postprocess::SampleDesc<Out, In>;
+
+  ToFreqMajorSpectrum() = default;
+  explicit ToFreqMajorSpectrum(Convert convert) : convert_(convert) {}
+
+  KernelRequirements Setup(KernelContext &ctx, const TensorListShape<2> &in_shape) override {
+    KernelRequirements req;
+    ScratchpadEstimator se;
+    req.output_shapes.resize(1);
+    auto &out_shape = req.output_shapes[0];
+
+    int N = in_shape.num_samples();
+    if (N == 0) {
+      nblocks_ = 0;
+      return req;
+    }
+    int nblocks = 0;
+    int nfft = in_shape[0][1];
+
+    int64_t total_windows = 0;
+    out_shape.resize(N, 2);
+    for (int i = 0; i < N; i++) {
+      TensorShape<2> sample_shape = in_shape[i];
+      DALI_ENFORCE(sample_shape[1] == nfft, "All inputs must have the same number of FFT bins");
+      out_shape.set_tensor_shape(i, { sample_shape[1], sample_shape[0] });
+
+      total_windows += sample_shape[0];
+    }
+
+    // if the data is big, use larger blocks
+    block_size_ = kBlock * (1 + total_windows / (100000 * kBlock));
+
+    for (int i = 0; i < N; i++) {
+      int nwindows = in_shape[i][0];
+      nblocks += div_ceil(nwindows, block_size_);
+    }
+    nblocks_ = nblocks;
+    se.add<SampleDesc>(AllocType::GPU, N);
+    se.add<BlockDesc>(AllocType::GPU, nblocks);
+    se.add<SampleDesc>(AllocType::Host, N);
+    se.add<BlockDesc>(AllocType::Host, nblocks);
+    req.scratch_sizes = se.sizes;
+    return req;
+  }
+
+  void Run(KernelContext &ctx, const OutListGPU<Out, 2> &out, const InListGPU<In, 2> &in) override {
+    int N = in.num_samples();
+    if (!N)
+      return;
```

Comment:
How other operator would behave if it return empty output.

Reply:
If they fail, they should be fixed. While one can argue about empty recordings/images, there's nothing fundamentally wrong with an empty list thereof.

Reply:
I think our operators and FW iterators may fail.